pkg/odo/cli/catalog/list/components.go

```diff
@@ -54,35 +57,39 @@ func (o *ListComponentsOptions) Validate() (err error) {
 
 // Run contains the logic for the command associated with ListComponentsOptions
 func (o *ListComponentsOptions) Run() (err error) {
-	w := tabwriter.NewWriter(os.Stdout, 5, 2, 3, ' ', tabwriter.TabIndent)
-	var supCatalogList, unsupCatalogList []catalog.CatalogImage
+	if log.IsJSON() {
+		machineoutput.OutputSuccess(o.catalogList)
+	} else {
+		w := tabwriter.NewWriter(os.Stdout, 5, 2, 3, ' ', tabwriter.TabIndent)
+		var supCatalogList, unsupCatalogList []catalog.ComponentType
+
+		for _, image := range o.catalogList.Items {
+			supported, unsupported := catalog.SliceSupportedTags(image)
+
+			if len(supported) != 0 {
+				image.Spec.NonHiddenTags = supported
+				supCatalogList = append(supCatalogList, image)
+			}
+			if len(unsupported) != 0 {
+				image.Spec.NonHiddenTags = unsupported
+				unsupCatalogList = append(unsupCatalogList, image)
+			}
+		}
 
-	for _, image := range o.catalogList {
-		supported, unsupported := catalog.SliceSupportedTags(image)
+		if len(supCatalogList) != 0 {
+			fmt.Fprintln(w, "Odo Supported OpenShift Components:")
+			o.printCatalogList(w, supCatalogList)
+			fmt.Fprintln(w)
 
-		if len(supported) != 0 {
-			image.NonHiddenTags = supported
-			supCatalogList = append(supCatalogList, image)
-		}
-		if len(unsupported) != 0 {
-			image.NonHiddenTags = unsupported
-			unsupCatalogList = append(unsupCatalogList, image)
 		}
-	}
-
-	if len(supCatalogList) != 0 {
-		fmt.Fprintln(w, "Odo Supported OpenShift Components:")
-		o.printCatalogList(w, supCatalogList)
-		fmt.Fprintln(w)
 
-	}
+		if len(unsupCatalogList) != 0 {
+			fmt.Fprintln(w, "OpenShift Components:")
+			o.printCatalogList(w, unsupCatalogList)
+		}
 
-	if len(unsupCatalogList) != 0 {
-		fmt.Fprintln(w, "OpenShift Components:")
-		o.printCatalogList(w, unsupCatalogList)
+		w.Flush()
```

Comment:
Is this output expected:
```sh
$ ./odo catalog list components
OpenShift Components:
NAME        PROJECT       TAGS
dotnet      openshift     2.0,latest
httpd       openshift     2.4,latest
nginx       openshift     1.10,1.12,1.8,latest
nodejs      openshift     10,6,8,8-RHOAR,latest
perl        openshift     5.24,5.26,latest
php         openshift     7.0,7.1,latest
python      openshift     2.7,3.5,3.6,latest
ruby        openshift     2.3,2.4,2.5,latest
wildfly     openshift     10.0,10.1,11.0,12.0,13.0,8.1,9.0,latest
```
I was expecting list of supported & unsupported catalog components.

Reply:
this is because of https://github.com/openshift/odo/issues/2129. you can use the PSI cluster which should show you supported and unsupported

Reply:
Why do we need to check aginst PSI cluster? If I check against any other cluster, it won't show what's supported and what's not? :thinking:

Reply:
upstream images vs downstream images. and we have PR which will resolve it - https://github.com/openshift/odo/issues/2129

Reply:
upstream images have different repositories and do observe that the above list doesn't even mention `java`, this is because due to licensing upstream okd cannot have java shipped in it.
And as per nodejs images, the repository urls for component image streams are different between upstream and downstream openshift. 
We added the downstream images in `supported` list. but forgot to add the `upstream` images.

Reply:
which will soon be resolved with this https://github.com/openshift/odo/pull/2140

Reply:
Yet, doing `odo catalog list components` should list them as "Unsupported", right?

Reply:
> Yet, doing `odo catalog list components` should list them as "Unsupported", right?

yes, that is right. without #2140 you should see all images as unsupported. So there is still something wrong with the output

Reply:
not really, unless the PR isn't merged this should be the expected output

Reply:
What is @dharmit trying to point out is that if someone has a cluster with no supported image (only with unsupported once) The title should not be just `OpenShift Components:`, but `OpenShift UNSUPPORTED Components:`
It is in related to #2140, as even once this is merged the will be still present for someone who doesn't have a single supported image on the cluster.

Reply:
I disagree. if all the components are not fully supported then we shouldn't say that `Openshift UNSUPPORTED Components` it gives a wrong message.

Reply:
`Unsupported` is an ambiguous term which could be taken by the user as that they cannot even use odo with their cluster.
The current output in my opinion effective and actually adapts to this scenario well.

Reply:
also this is not in scope of this PR and hence can be done on a separate issue related to "improving catalog component output"
@dharmit WDYT?

Reply:
removing the hold

Reply:
> Unsupported is an ambiguous term which could be taken by the user as that they cannot even use odo with their cluster.
The current output in my opinion effective and actually adapts to this scenario well.


How we can say that on one cluster we officially support the image, and on the other one we don't?

Reply:
>also this is not in scope of this PR and hence can be done on a separate issue related to "improving catalog component output"
@dharmit WDYT?

Yes, but this is still an issue that needs to be fixed. @dharmit can you open a new issue for this?

Reply:
#2148